zkevm-circuits/src/bin/stats/halo2_stats.rs

```diff
@@ -0,0 +1,373 @@
+    //
+    //
+    // ### 7. Turn advice_lagranges into advice_monomials.
+    // ```
+    // let advice_monomials = advice_lagranges.to_monomials();
+    // drop(advice_lagranges);
+    // ```
+    //
+    //
+    // ### 8. Generate necessary extended lagranges.
+    //
+    // M.C = 2
+    // M.S = 1 + 2 * c_f + 2 * c_p + 2 * c_i + c_a + 3 * c_l + c_pg
+    // M.E = 3 + c_f + c_p + c_pg + c_i + c_a (+= c_i + c_a)
+    // ```
+    // let instances_extended_lagrnages = instances_monomials.to_extended_lagranges();
+    // let advice_extended_lagrnages = advice_monomials.to_extended_lagranges();
+    // ```
+    //
+    //
+    // ### 9. While computing the quotient, these extended lagranges:
+    //
+    // ⬡ permuted_input_extended_lagranges
+    // ⬡ permuted_table_extended_lagranges
+    // ⬡ lookup_product_extended_lagranges
+    //
+    // of each lookup argument are generated on the fly and drop before next.
+    //
+    // And 1 extra quotient_extended_lagrange is created. So the peak memory:
+    //
+    // M.C = 2
+    // M.S = 1 + 2 * c_f + 2 * c_p + 2 * c_i + c_a + 3 * c_l + c_pg
+    // M.E = 4 + c_f + c_p + c_pg + c_i + c_a + 3 * (c_l > 0) (+= 3 * (c_l > 0) + 1)
+    // ```
+    // let quotient_extended_lagrange = quotient_extended_lagrange();
+    // ```
+    //
+    //
+    // ### 10. After quotient is comuputed, drop all the other extended lagranges.
+    //
+    // M.C = 2
+    // M.S = 1 + 2 * c_f + 2 * c_p + 2 * c_i + c_a + 3 * c_l + c_pg
+    // M.E = 4 + c_f + c_p (-= c_pg + c_i + c_a + 3 * (c_l > 0))
+    // drop(instances_extended_lagrnages)
+    // drop(advice_extended_lagrnages)
+    // drop(perm_grand_product_extended_lagranges)
+    //
+    //
+    // ### 11. Turn quotient_extended_lagrange into monomial form.
+    // And then cut int `d - 1` pieces.
+    //
+    // M.C = 2
+    // M.S = 2 * c_f + 2 * c_p + 2 * c_i + c_a + 3 * c_l + c_pg + d (+= d - 1)
+    // M.E = 3 + c_f + c_p (-= 1)
+    // ```
+    // let quotient_monomials = quotient_monomials()
+    // drop(quotient_extended_lagrange)
+    // ```
+    //
+    //
+    // ### 12. Evaluate and open all polynomial except instance ones.
+    // }
+    pub(crate) fn estimate_peak_mem(&self, k: u32) -> usize {
+        let field_bytes = 32;
+        let c_f = self.num_fixed_columns;
+        let c_a = self.num_advice_columns;
+        let c_i = self.num_instance_columns;
+        let c_p = self.num_permutation_columns;
+        let c_l = self.num_lookups;
+        let c_pg = (c_p + self.degree - 3) / (self.degree - 2);
+        let e = (self.degree - 1).next_power_of_two();
+        // number of "n elliptic curve points"
+        let m_c = 2;
+        // number of "n field elements"
+        let m_s = 1 + 2 * c_f + 2 * c_p + 2 * c_i + c_a + 3 * c_l + c_pg;
+        // number of "e * n field elements"
+        let m_e = 4 + c_f + c_p + c_pg + c_i + c_a + 3 * (c_l > 0) as usize;
+        let unit = m_c + m_s + e * m_e;
+        unit * 2usize.pow(k) * field_bytes
```

Comment:
`unit * 2usize.pow(k) * field_bytes` looks like to calculate the memory usage as above formula `M = 32 * n * (2 * M.C + M.S + e * M.E)` for me. But it seems not exact the same.

The implementation here is
unit * 2usize.pow(k) * field_bytes --> (m_c + m_s + e * m_e) * n * 32
however M is
(`2 * M.C` + M.S + e * M.E) * n * 32

Wondering which one is correct?

Reply:
Yes you're correct, the `2 *` is missing!  Thanks for pointing this out!  I also saw this when checking your review message https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1763#discussion_r1493946938 :smiley: